docs/2019-04-23-Applicative-Kore.md

```diff
@@ -1,222 +1,286 @@
+| etc.              | etc.                          | etc.                          |
+
+_Remark._ The attributes `left-assoc` and `right-assoc` do not carry semantic meaning. They carry only syntactic meaning. Specifically speaking, they tell the _verifier_ to accept patterns of the form `foo a1 a2 a3`, even if `foo` has been declared to have arity 2. Without these attributes, the verifier will reject the pattern `foo a1 a2 a3` because it is not wellformed.
+
+As an example, here is the declaration of `\or` using `left-assoc` attribute:
+
+```
+notation \or P Q = \implies (\neg P) Q [left-assoc]
 ```
-notation \and P Q = \neg (\or (\neg P) (\neg Q))
-left-assoc-notation \and
+
+Another neat example is the declaration of `\exists` that allows multiple binders:
+
 ```
+notation \exists X P = \neg (\forall X (\neg P)) [binder, right-assoc]
+```
+
+Here, the `right-assoc` attribute tells the verifier to accept patterns of the form `\exists X1 X2 ... Xn P` as a notation of the pattern `(\exists X1 ... (\exists Xn P) ...)`.
 
-Volatile `\forall` can be defined as
+___QUESTION.___ How to do the same thing for `\forall`? Note that unlike `\exists`, `\forall` is one of the primitive construct that takes exactly one binder, as specified by the AppKore syntax. Can we declare:
 
 ```
-right-assoc-notation \forall
+notation \forall X P = \forall X P [binder, right-assoc]  // try to make \forall valatile
 ```
 
-which tells the verifier to accept patterns of the form `\forall X1 X2 ... Xn P` and desugar it (if required), uniquely, to `(\forall X1 ... (\forall Xn P) ...)`.
+Note that `\forall` now also appears on the right hand side. This is not intended, because we want to forbid recursive notations. 
 
+#### Hooked notations
 
+Some notations are not definable in the logic. One such examples is _capture-avoiding substitution_. Specifically, we want to define `\subst X P Q` as a notation that represents the pattern obtained by substituting all free occurrences of `X` in `P` for `Q`, where alpha-renaming happens explicitly to prevent unintended variable capturing. 
+
+To achieve this, we use _hooked notations_. Hooked notations are notations that do not have an explicit definition in their declarations. Instead, the backends should provide proper implementation of hooked notations. 
+
+As an example, `\subst` can be defined as follows:
 
 ```
-notation \neg P = \implies P \bot                             [basic]
-notation \iff P Q = \and (\implies P Q) (\implies Q P)        [basic]
-notation \top = \neg \bot                                     [basic]
-notation \exists x P = \neg \forall x \neg P                  [basic]
-notation \nu X P = \neg \mu X (\neg (\subst P (\neg X) X))    [basic]
-notation \and P Q ... = \and2 P (\and Q ...)                  [basic]
-notation \or P Q ... = \or2 P (\or Q ...)                     [basic]
+hooked-notation \subst X P Q [required, desugar]
```

Comment:
Because this notation is `hooked-`, isn't `required` a redundant attribute?

Reply:
I think maybe we'll need some meta-notations (i.e., hooked-notations) for one purpose (e.g., verification), in which case other backends (e.g., those only for execution) won't need it. Substitution, however, is essential to all backends, and thus its implementation is required and forced in all backends.